docs/core_docs/docs/tutorials/qa_chat_history.ipynb

```diff
@@ -1,1438 +1,943 @@
+    "  // Run\n",
+    "  const response = await llm.invoke(prompt)\n",
+    "  return { messages: [response] };\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b409ee5f-2973-47ee-a1bf-112731843c5d",
+   "metadata": {},
+   "source": [
+    "Finally, we compile our application into a single `graph` object. In this case, we are just connecting the steps into a sequence. We also allow the first `query_or_respond` step to \"short-circuit\" and respond directly to the user if it does not generate a tool call. This allows our application to support conversational experiences-- e.g., responding to generic greetings that may not require a retrieval step"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "id": "60ed5dd0-8636-4a12-961b-b8355b579862",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import { StateGraph } from \"@langchain/langgraph\";\n",
+    "import { toolsCondition } from \"@langchain/langgraph/prebuilt\";\n",
+    "\n",
+    "\n",
+    "const graphBuilder = new StateGraph(MessagesAnnotation)\n",
+    "  .addNode(\"queryOrRespond\", queryOrRespond)\n",
+    "  .addNode(\"tools\", tools)\n",
+    "  .addNode(\"generate\", generate)\n",
+    "  .addEdge(\"__start__\", \"queryOrRespond\")\n",
+    "  .addConditionalEdges(\n",
+    "    \"queryOrRespond\",\n",
+    "    toolsCondition,\n",
+    "    {__end__: \"__end__\", tools: \"tools\"}\n",
+    "  )\n",
+    "  .addEdge(\"tools\", \"generate\")\n",
+    "  .addEdge(\"generate\", \"__end__\")\n",
+    "\n",
+    "const graph = graphBuilder.compile();"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "0a0389e0-c454-4693-b6e2-979326015574",
+   "metadata": {},
+   "source": [
+    "```javascript\n",
+    "// Note: tslab only works inside a jupyter notebook. Don't worry about running this code yourself!\n",
+    "import * as tslab from \"tslab\";\n",
+    "\n",
+    "const image = await graph.getGraph().drawMermaidPng();\n",
+    "const arrayBuffer = await image.arrayBuffer();\n",
+    "\n",
+    "await tslab.display.png(new Uint8Array(arrayBuffer));\n",
+    "```\n",
+    "\n",
+    "![graph_img_rag_part_2](../../static/img/graph_img_rag_part_2.png)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3aee3fb6-c7a2-44a5-96f0-2951631caaf2",
+   "metadata": {},
+   "source": [
+    "Let's test our application.\n",
+    "\n",
+    "```{=mdx}\n",
+    "<details>\n",
+    "<summary>Expand for `prettyPrint` code.</summary>\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "id": "cb29afc4-899b-4f11-89bf-092368d21bec",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import { BaseMessage, isAIMessage } from \"@langchain/core/messages\";\n",
+    "\n",
+    "const prettyPrint = (message: BaseMessage) => {\n",
+    "  let txt = `[${message._getType()}]: ${message.content}`;\n",
+    "  if (\n",
+    "    (isAIMessage(message) && (message as AIMessage)?.tool_calls?.length) ||\n",
```

Comment:
You shouldn't need the casts after `isAIMessage`

https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards